### DISCUSSION TITLE
Query Scopes: manage them for queued Jobs, controller routes and relationships

### DISCUSSION THREAD

**Discussion by didac-adria** [laravel/framework#51536] [Ideas]
Hey Laravel Community!

This is a bit of an idea, debate and open question. Recently we have implemented [Query Scopes](https://laravel.com/docs/10.x/eloquent#query-scopes) in our project. We manage Bookings for Rentals. We want certain type of booking Statuses to be out of all our queries, for instance, Cancelations or Requests to Book (not yet confirmed reservations). So we implemented the above mentioned query scopes.

Now, sometimes you have to do things with bookings which are oursie of the scope, like with canceled bookings. I may create a job to handle the things to do after canceling a reservation in the database. I pass this to a job in a queue and yeah, obviously it doesn't work. Why? Because Laravel serializes the model by id, and when retrieving back such model, uses the query scopes and canceled bookings are outside of the scope, thus the job fails. The solution I've found is to pass the id and we do our own query of the booking.

However, this is a recurring problem because happens exactly the same with controller [routes with model binding](https://laravel.com/docs/10.x/routing#route-model-binding) and with relationships. In both cases, you are looking for a particular booking, regardless of it's state. So if I'm looking for the route /bookings/124, I'm looking for such booking even though it's canceled. If I'm within a booking payment and I'm looking to use the relationship to get the parent booking ($bookingPayment->booking), even though it's canceled, I wish to find it. Now both problems have solutions, but they are a bit outside of the Laravel standards and decrease maintainability. For routes it's nicer, I've implemented the resolveRouteBindingQuery() method inside the Laravel Model class, but for routes, I have to tell Laravel that all relationships pointing to booking must be without any query scope.

Now opening up the debate, should we add to Laravel something where you can define de behaviour of each Query Scope? Just as an idea, some methods to define the default behaviour for routes, relationships and jobs? For instance something like this:

```
namespace App\Models\Scopes;

class WithoutCancelationsScope implements Scope
{
    /**
     * Apply the scope to a given Eloquent query builder.
     */
    public function apply(Builder $builder, Model $model): void
    {
        $builder->where('status', '!=', BookingStatusEnum::CANCELED)
            ->where('canceled_at', 0);
    }
    
    public function activeForRoutes(): bool
    {
        return false;
    }

    public function activeForJobs(): bool
    {
        return false;
    }

    public function activeForRelationships(): bool
    {
        return false;
    }
}

```

Currently Laravel uses Scopes to handle Soft deletes. For routes we have a solution to query deleted models, but I think there is no solution to pass a soft deleted model to a Job. So would be useful for such cases as well.

Regards!

**Comment by macropay-solutions** [comment]
Interesting suggestion. 
We solved the routes issue with an external free lib that is not affected by scopes and replaced the scopes with filters.
The relation's scopes we solved by defining a relation for each scope if ever needed, or by filtering that relation to restrict the resource results.

For queues we never serialize models but their ids and we query for them in the job.

We did not see the usefulness of scopes.

But now that you posted this and need to force the scope in the route (not like the optional query filter):

list all (without scopes) with filters:
GET /all/resource

list not deleted with filters:
GET /info/resource

list deleted with filters:
GET /deleted/resource


so in conclusion the scope could be introduced in route as a prefix.

GET /{scope}/resource

**Reply by didac-adria** [reply]
Nice to see how you guys solved it! Yes, there are solutions but they are all ugly and not well maintainable, which is the purpose of Laravel framework. For instance, we have to remember that each time we pass a booking to a job, it must be in the form of an id and not a model. Otherwise this could cause not easy to track issues.

Let's see if other community members respond and see the interest of it. I understand that if there is sufficient interest in the community, this could be interesting for future Laravel releases.

**Reply by macropay-solutions** [reply]
We call those internal rules. Laravel has many things built in that we avoid using just because unpredictable things can happen.

For example.

- Any sql query can throw an error, so leaving laravel deal with that is not ok for us. We prefer to fetch the data from db "in sight" not behind the scenes when it comes to jobs and models.
- The built in apiResource.
-  and so on.
